src/inmanta/data/dataview.py

```diff
@@ -0,0 +1,1147 @@
+        result = cast(List[Record], result)
+        if not result:
+            raise InvalidQueryParameter("Could not determine page bounds")
+        return PagingMetadata(
+            total=cast(int, result[0]["count_total"]),
+            before=cast(int, result[0].get("count_before", 0)),
+            after=cast(int, result[0].get("count_after", 0)),
+            page_size=self.limit,
+        )
+
+    async def prepare_paging_links(
+        self,
+        dtos: Sequence[T_DTO],
+        paging_boundaries: Union[PagingBoundaries, RequestedPagingBoundaries],
+        meta: PagingMetadata,
+    ) -> Dict[str, str]:
+        """
+        Construct the paging links
+        """
+        links = {}
+
+        url_query_params: Dict[str, Optional[Union[SimpleTypes, List[str]]]] = {
+            "limit": self.limit,
+            "sort": str(self.order),
+        }
+
+        for key, value in self.raw_filter.items():
+            if value is not None:
+                url_query_params[f"filter.{key}"] = value
+
+        url_query_params.update(self.get_extra_url_parameters())
+
+        if dtos:
+            base_url = self.get_base_url()
+
+            def value_to_string(value: Union[str, int, UUID, datetime]) -> str:
+                if isinstance(value, datetime):
+                    # Accross API boundaries, all naive datetime instances are assumed UTC.
+                    # Returns ISO timestamp implicitly in UTC.
+                    return datetime_utc_isoformat(value, naive_utc=True)
+                return str(value)
+
+            def make_link(**args: Optional[Union[str, int, UUID, datetime]]) -> str:
+                params = url_query_params.copy()
+                params.update({k: value_to_string(v) for k, v in args.items() if v is not None})
+                return f"{base_url}?{parse.urlencode(params, doseq=True)}"
+
+            link_with_end = make_link(
+                end=paging_boundaries.end,
+                last_id=paging_boundaries.last_id,
+            )
+            link_with_start = make_link(
+                start=paging_boundaries.start,
+                first_id=paging_boundaries.first_id,
+            )
+
+            has_next = meta.after > 0
+            if has_next:
+                if self.order.get_order() == "DESC":
+                    links["next"] = link_with_end
+                else:
+                    links["next"] = link_with_start
+
+            has_prev = meta.before > 0
+            if has_prev:
+                if self.order.get_order() == "DESC":
+                    links["prev"] = link_with_start
+                else:
+                    links["prev"] = link_with_end
+                # First page
+                links["first"] = make_link()
+
+            # Same page
+            links["self"] = make_link(
+                first_id=self.requested_page_boundaries.first_id,
+                start=self.requested_page_boundaries.start,
+            )
+            # TODO: last links
```

Comment:
Another TODO here

Reply:
It is a real todo, for one day, in the future, if someone has the courage